File: scripts/fetch_prices.mjs
// scripts/fetch_prices.mjs
import fs from "fs";

const ENTRY_DATE = "2026-02-11"; // YYYY-MM-DD (Feb 11, 2026)
const OUT_PATH = "data/prices.json";

// Adjust these if any symbol fails (errors will be written to prices.json)
const TICKERS = [
  { key: "AGI",  stooq: "AGI.US"  },
  { key: "FSM",  stooq: "FSM.US"  },
  { key: "GAU",  stooq: "GAU.US"  },
  { key: "NFGC", stooq: "NFGC.US" },
  { key: "VGZ",  stooq: "VGZ.US"  },
  { key: "NEWP", stooq: "NEWP.US" }
];

function pickTag(xml, tag) {
  const m = xml.match(new RegExp(`<${tag}>(.*?)</${tag}>`, "i"));
  return m ? m[1].trim() : null;
}

async function fetchQuoteXML(symbol) {
  const url = `https://stooq.pl/q/l/?s=${encodeURIComponent(symbol)}&e=xml`;
  const res = await fetch(url, { headers: { "User-Agent": "marketbriefdaily-bot/1.0" } });
  if (!res.ok) throw new Error(`Quote HTTP ${res.status}`);
  const xml = await res.text();

  const date = pickTag(xml, "date");
  const time = pickTag(xml, "time");
  const close = pickTag(xml, "close");

  return {
    symbol,
    date,
    time,
    close: close ? Number(close) : null
  };
}

async function fetchEntryOpenFromCSV(symbol, entryDate) {
  // Stooq daily history CSV endpoint
  const url = `https://stooq.com/q/d/l/?s=${encodeURIComponent(symbol)}&i=d`;
  const res = await fetch(url, { headers: { "User-Agent": "marketbriefdaily-bot/1.0" } });
  if (!res.ok) throw new Error(`History HTTP ${res.status}`);

  const csv = await res.text();
  // CSV header: Date,Open,High,Low,Close,Volume
  const lines = csv.trim().split(/\r?\n/);
  if (lines.length < 2) throw new Error("History empty");

  // Find row matching ENTRY_DATE
  // Example line: 2026-02-11,12.34,12.50,12.20,12.45,123456
  for (let i = 1; i < lines.length; i++) {
    const parts = lines[i].split(",");
    if (parts[0] === entryDate) {
      const open = Number(parts[1]);
      if (!Number.isFinite(open)) throw new Error("Bad open value");
      return open;
    }
  }

  throw new Error(`No row for ${entryDate} (market holiday or symbol mismatch)`);
}

async function main() {
  const quotes = {};
  const errors = [];

  for (const t of TICKERS) {
    try {
      const [q, entryOpen] = await Promise.all([
        fetchQuoteXML(t.stooq),
        fetchEntryOpenFromCSV(t.stooq, ENTRY_DATE)
      ]);

      quotes[t.key] = {
        symbol: t.stooq,
        entryDate: ENTRY_DATE,
        entryOpen,
        last: q.close,     // latest close/quote Stooq provides
        quoteDate: q.date,
        quoteTime: q.time
      };
    } catch (e) {
      errors.push({ ticker: t.key, symbol: t.stooq, error: String(e.message || e) });
    }
  }

  const out = {
    asOf: new Date().toISOString(),
    entryDate: ENTRY_DATE,
    source: "stooq",
    quotes,
    errors: errors.length ? errors : undefined
  };

  fs.writeFileSync(OUT_PATH, JSON.stringify(out, null, 2));
  console.log(`Wrote ${OUT_PATH}`);
  if (errors.length) console.warn("Some tickers failed:", errors);
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});

File: scripts/fetch_prices.py
import json
import time
import urllib.request
from datetime import datetime, timezone

# Stooq daily snapshot CSV:
# https://stooq.com/q/l/?s=agi.us&f=sd2t2ohlcv&h&e=csv
# fields: s = symbol, d2 = date, t2 = time, o = open, h = high, l = low, c = close, v = volume

TICKERS = {
    "AGI": "agi.us",
    "FSM": "fsm.us",
    "GAU": "gau.us",
    "NFGC": "nfgc.us",
    "VGZ": "vgz.us",
    "NEWP": "newp.us",
    "GOOGL": "googl.us",
}

def fetch_csv(symbol: str) -> str:
    url = f"https://stooq.com/q/l/?s={symbol}&f=sd2t2ohlcv&h&e=csv"
    with urllib.request.urlopen(url, timeout=20) as r:
        return r.read().decode("utf-8", errors="replace")

def parse_latest(csv_text: str):
    lines = [ln.strip() for ln in csv_text.splitlines() if ln.strip()]
    if len(lines) < 2:
        return None
    header = lines[0].split(",")
    row = lines[1].split(",")
    if len(row) != len(header):
        return None

    data = dict(zip(header, row))
    # Expect columns: Symbol,Date,Time,Open,High,Low,Close,Volume
    # Sometimes Stooq returns "N/A" values
    def fnum(x):
        try:
            return float(x)
        except:
            return None

    return {
        "date": data.get("Date"),
        "time": data.get("Time"),
        "open": fnum(data.get("Open")),
        "price": fnum(data.get("Close")),  # "Close" as current snapshot
    }

def main():
    out = {
        "asOf": datetime.now(timezone.utc).isoformat(timespec="seconds"),
        "source": "stooq",
        "quotes": {}
    }

    for tkr, stooq_symbol in TICKERS.items():
        try:
            csv_text = fetch_csv(stooq_symbol)
            q = parse_latest(csv_text)
            if q and (q["open"] is not None or q["price"] is not None):
                out["quotes"][tkr] = {"open": q["open"], "price": q["price"]}
            else:
                out["quotes"][tkr] = {"open": None, "price": None}
        except Exception:
            out["quotes"][tkr] = {"open": None, "price": None}

        time.sleep(0.8)  # be polite

    with open("data/prices.json", "w", encoding="utf-8") as f:
        json.dump(out, f, indent=2)

if __name__ == "__main__":
    main()
